### ISSUE TITLE
Design an new `Image` class for receive images with different sizes.

### ISSUE THREAD

**Issue by su-chang** [wechaty/wechaty#1871]
**Is your feature request related to a problem? Please describe.**

We can download image by this method `toFileBox()`, but we can not choice the image quality. Sometime we need compressed image only.

**Describe the solution you'd like**

Maybe we could set a param for `toFileBox()`, just like this way : `toFileBox(compressed: boolean): FileBox`, and the default param value is `true`.

**Describe alternatives you've considered**

**Additional context**

If you agree with me, I will create a PR for it.

[enhancement]

**Comment by huan** [comment]
What if the message instance is not an image, let's say, it's an attachment file like a PDF?

**Comment by su-chang** [comment]
Good catch! Maybe we should ignore this options in other cases except image.

**Comment by huan** [comment]
Sorry, that will not be acceptable, because it will make the API dirty. We should make the API clean as possible as we can.

The thumb option is totally non business related to the FileBox method, so we should avoid to put it here.

I'd like to think about to add a new method like `toImage()` which return a simple WeChat Image object, it can be convert to thumb, or the full sized image.

Could you please list all the thumb related scanerioes so that we can see it clearer before any further discussion from us?

**Comment by su-chang** [comment]
Thanks for your suggestion, I agree with you.

- raw payload message for image
```js
{
  "content": "<?xml version=\"1.0\"?>\n<msg>\n\t<img aeskey=\"f880e5baa73ef65daf87626f7d2d9304\" encryver=\"1\" cdnthumbaeskey=\"f880e5baa73ef65daf87626f7d2d9304\" cdnthumburl=\"304f0201000448304602010002046212e73902033d0af802042f30feb602045dc4c2e40421777869645f65326c63333361757536316732323130345f313537333137363033350204010400020201000400\" cdnthumblength=\"26682\" cdnthumbheight=\"120\" cdnthumbwidth=\"120\" cdnmidheight=\"0\" cdnmidwidth=\"0\" cdnhdheight=\"0\" cdnhdwidth=\"0\" cdnmidimgurl=\"304f0201000448304602010002046212e73902033d0af802042f30feb602045dc4c2e40421777869645f65326c63333361757536316732323130345f313537333137363033350204010400020201000400\" length=\"29472\" md5=\"b54982daab2bdbf7fa5a2e1d5d7f3451\" />\n</msg>\n",
  "createTime": 1573176144267,
  "fromUserName": "Soul001001",
  "imgStatus": 2,
  "l1MsgType": 5,
  "msgId": "5705251123828180997",
  "msgSource": "<msgsource>\n\t<img_file_name>1051573176083_.pic.jpg</img_file_name>\n</msgsource>\n",
  "msgSourceCd": 2,
  "msgType": 3,
  "newMsgId": 5705251123828181000,
  "pushContent": " abc : [图片]",
  "status": 3,
  "toUserName": "wxid_e2lc33auu61g22",
  "uin": "3710862731",
  "url": "https://qiniu2.lianluoyi.cn/msg/CJ3FC/20191108933627732/5705251123828180997_wxid_e2lc33auu61g22_1573176140135_.png",
  "wechatUserName": "wxid_e2lc33auu61g22"
}
```
- raw payload in Padplus
```js
{
  "id": "5705251123828180997",
  "timestamp": 1573176144.267,
  "type": 6,
  "filename": "5705251123828180997.jpg",
  "fromId": "Soul001001",
  "text": "<?xml version=\"1.0\"?>\n<msg>\n\t<img aeskey=\"f880e5baa73ef65daf87626f7d2d9304\" encryver=\"1\" cdnthumbaeskey=\"f880e5baa73ef65daf87626f7d2d9304\" cdnthumburl=\"304f0201000448304602010002046212e73902033d0af802042f30feb602045dc4c2e40421777869645f65326c63333361757536316732323130345f313537333137363033350204010400020201000400\" cdnthumblength=\"26682\" cdnthumbheight=\"120\" cdnthumbwidth=\"120\" cdnmidheight=\"0\" cdnmidwidth=\"0\" cdnhdheight=\"0\" cdnhdwidth=\"0\" cdnmidimgurl=\"304f0201000448304602010002046212e73902033d0af802042f30feb602045dc4c2e40421777869645f65326c63333361757536316732323130345f313537333137363033350204010400020201000400\" length=\"29472\" md5=\"b54982daab2bdbf7fa5a2e1d5d7f3451\" />\n</msg>\n",
  "url": "https://qiniu2.lianluoyi.cn/msg/CJ3FC/20191108933627732/5705251123828180997_wxid_e2lc33auu61g22_1573176140135_.png",
  "toId": "wxid_e2lc33auu61g22"
}
```
- load the full sized image for the image message
```js
{
  "msgId": "5705251123828180997",
  "src": "https://qiniu2.lianluoyi.cn/msg/CJ3FC/201911081423139848/5705251123828180997_wxid_e2lc33auu61g22_1573176140854_.png",
  "thumb": "https://qiniu2.lianluoyi.cn/msg/CJ3FC/20191108933627732/5705251123828180997_wxid_e2lc33auu61g22_1573176140135_.png"
}
```

the `src` is the full sized image url.

**Comment by windmemory** [comment]
Is there any progress on this thread? @huan @su-chang

**Comment by huan** [comment]
@windmemory more draft design from @su-chang  will be welcome!

**Comment by huan** [comment]
For example, if we abstract a new class named `XXXImage`:

1. What candadites will be able to descript this new class? Why?
1. What methods does this new class need?

**Comment by su-chang** [comment]
The new class named `ImageBox`, similar to `FileBox`. Maybe it could be accepted and understood easily.

And we have to add a new method for `Message` named `toImageBox()`, and will return `FileBox` for user to manipulate it.

Each Image message could be a ImageBox, it has two methods like:

```js
class ImageBox {
  thumbUrl: string, // extracted from Image message
  thumbImage(): FileBox // return the thumbUrl
  srcImage(): Promise<FileBox> // request for full size image url
}
```

### Example
```js
if (message.type() === MessageType.Image) {
  const imageBox: ImageBox = await message.toImageBox()
  const thumbFileBox = imageBox.thumbImage()
  const srcFileBox = await imageBox.srcImage()
}
```

So `thumbImage()` will be the basic method, and if someone want to load full size image, and then just call `srcImage()`.

@huan How do you think about it? Waiting for your suggestions. Thank you.

**Comment by huan** [comment]
Thanks for the design of the image abstract class. Will look into it later.

**Comment by su-chang** [comment]
@huan ping

**Comment by su-chang** [comment]
@huan ping

**Comment by huan** [comment]
How about we create a new class, say, `class Image`, which will be returned from `Message.toImage()`:

```ts
class Image {
  thumbnail(): FileBox
  hd(): FileBox
  artwork(): FileBox
}
```

Just as the name tells us: `thumbnail()` is the smallest, `hd()` is the normal size, and the `artwork()` will be the original resolution.

I put three methods instead of two at here, is because I remember that the wechat image has three format:

1. in chat conversation interface, is the thumbnail
1. if we click the image to view it, the resolution is getting better
1. if we click the original button to view it, the resolution is the best.

What do you think about the design? 

Please feel free to let me know your idea, and if you think it's ok, could you please contribute some design for puppet API, like `messageImageThumbnail()`, `messageImageHd()`, and `messageImageArtwork()`?

**Comment by windmemory** [comment]
Can we also add a method to get the url of the image? If the url is able to be provided by the underlying puppet, then just get the url will be definitely save a lot of network traffic.

**Comment by huan** [comment]
Of course, we can add a url method to let the user get the URL.

I did not add it because I remember that as the protocol of the web version, image url can not be visited without the cookie. It may differ between the protocol, so I'm ok with that.

However, I believe the FileBox solution will be the best practice, which means it will always work no matter what protocol we are using.

**Comment by windmemory** [comment]
I agree, FileBox should be the best practice. `url` is just for some specific cases.

**Comment by huan** [comment]
As I just re-designed the `Image` class, I  feel that the `url()` is not necessary because:

##### We can use `FileBox.fromUrl(url)` to archive that.

```ts
const fileBox = image.hd()
// the fileBox is generated from `FileBox.fromUrl(url)`
```

**Comment by huan** [comment]
It should be available with Wechaty@0.31.17 or above versions.
